File: client_app/invoicing/pdf/templates/classic.py
"""Classic tax-invoice PDF layout (ReportLab)."""

from decimal import Decimal
import os

from reportlab.lib import colors
from reportlab.lib.enums import TA_RIGHT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.platypus import Image, Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle

from invoicing.format import (
    format_abn,
    format_account,
    format_bsb,
    format_invoice_number,
    format_money,
    format_qty,
)
from invoicing.gst_settings import invoice_uses_tax_invoice

PAGE_WIDTH, PAGE_HEIGHT = A4
CONTENT_WIDTH = PAGE_WIDTH - 40 * mm

ACCENT = colors.HexColor("#2c3e50")
LIGHT_GREY = colors.HexColor("#f4f4f4")
BORDER_GREY = colors.HexColor("#cccccc")


def _p(text, style):
    return Paragraph(text.replace("\n", "<br/>"), style)


def _address_paragraphs(address, style):
    if not address:
        return []
    return [_p(line.strip(), style) for line in address.split("\n") if line.strip()]


def render_classic_invoice(output_dir, invoice_data):
    invoice_number = invoice_data["invoice_number"]
    invoice_number_display = format_invoice_number(invoice_number)
    invoice_date = invoice_data["invoice_date"]
    filename = f"Invoice_{invoice_number_display}_{invoice_date.isoformat()}.pdf"
    filepath = os.path.join(output_dir, filename)

    doc = SimpleDocTemplate(
        filepath,
        pagesize=A4,
        rightMargin=20 * mm,
        leftMargin=20 * mm,
        topMargin=20 * mm,
        bottomMargin=20 * mm,
    )

    styles = getSampleStyleSheet()
    business_name_style = ParagraphStyle(
        "BusinessName",
        parent=styles["Normal"],
        fontName="Helvetica-Bold",
        fontSize=14,
        leading=18,
        spaceAfter=4,
    )
    normal = ParagraphStyle(
        "NormalText",
        parent=styles["Normal"],
        fontSize=10,
        leading=14,
        spaceAfter=2,
    )
    label_style = ParagraphStyle(
        "Label",
        parent=normal,
        fontName="Helvetica-Bold",
        fontSize=9,
        textColor=colors.HexColor("#666666"),
        spaceAfter=4,
    )
    title_style = ParagraphStyle(
        "TaxInvoiceTitle",
        parent=styles["Normal"],
        fontName="Helvetica-Bold",
        fontSize=20,
        leading=26,
        alignment=TA_RIGHT,
        textColor=ACCENT,
        spaceAfter=0,
    )
    meta_style = ParagraphStyle(
        "Meta",
        parent=normal,
        alignment=TA_RIGHT,
        leading=14,
        spaceAfter=2,
        spaceBefore=0,
    )
    total_label_style = ParagraphStyle(
        "TotalLabel",
        parent=normal,
        alignment=TA_RIGHT,
        fontName="Helvetica-Bold",
        fontSize=10,
    )
    total_value_style = ParagraphStyle(
        "TotalValue",
        parent=normal,
        alignment=TA_RIGHT,
        fontSize=10,
    )
    grand_total_label = ParagraphStyle(
        "GrandTotalLabel",
        parent=total_label_style,
        fontSize=12,
    )
    grand_total_value = ParagraphStyle(
        "GrandTotalValue",
        parent=total_value_style,
        fontName="Helvetica-Bold",
        fontSize=12,
    )

    story = []

    gst_registered = invoice_uses_tax_invoice(invoice_data)
    invoice_title = "Tax Invoice" if gst_registered else "Invoice"

    meta_rows = [[Paragraph(invoice_title, title_style)]]
    meta_rows.append([Paragraph(f"<b>Invoice #</b> {invoice_number_display}", meta_style)])
    meta_rows.append([Paragraph(f"<b>Date</b> {invoice_date.strftime('%d %B %Y')}", meta_style)])
    due_date_fmt = invoice_data.get("due_date_fmt", "")
    due_rule_label = invoice_data.get("due_rule_label", "")
    payment_terms = invoice_data.get("payment_terms", "")
    if due_date_fmt:
        meta_rows.append([Paragraph(f"<b>Due</b> {due_date_fmt}", meta_style)])
    elif payment_terms:
        meta_rows.append([Paragraph(f"<b>Terms</b> {payment_terms}", meta_style)])

    meta_block = Table(meta_rows, colWidths=[CONTENT_WIDTH * 0.45])
    meta_block.setStyle(
        TableStyle(
            [
                ("ALIGN", (0, 0), (-1, -1), "RIGHT"),
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 0),
                ("RIGHTPADDING", (0, 0), (-1, -1), 0),
                ("TOPPADDING", (0, 0), (-1, -1), 0),
                ("BOTTOMPADDING", (0, 0), (0, 0), 10),
                ("BOTTOMPADDING", (0, 1), (-1, -1), 2),
            ]
        )
    )

    logo_block = [Paragraph("&nbsp;", normal)]
    if invoice_data.get("logo_enabled"):
        logo_path = invoice_data.get("logo_path")
        if logo_path and os.path.isfile(logo_path):
            try:
                img = Image(logo_path)
                max_w = 45 * mm
                max_h = 25 * mm
                iw, ih = float(img.imageWidth), float(img.imageHeight)
                scale = min(max_w / iw, max_h / ih, 1.0) if iw and ih else 1.0
                img.drawWidth = iw * scale
                img.drawHeight = ih * scale
                logo_block = [img]
            except Exception:
                logo_block = [Paragraph("&nbsp;", normal)]

    header_table = Table(
        [[logo_block, meta_block]],
        colWidths=[CONTENT_WIDTH * 0.55, CONTENT_WIDTH * 0.45],
    )
    header_table.setStyle(
        TableStyle(
            [
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 0),
                ("RIGHTPADDING", (0, 0), (-1, -1), 0),
            ]
        )
    )
    story.append(header_table)
    story.append(Spacer(1, 8))

    divider = Table([[""]], colWidths=[CONTENT_WIDTH], rowHeights=[2])
    divider.setStyle(TableStyle([("BACKGROUND", (0, 0), (-1, -1), ACCENT)]))
    story.append(divider)
    story.append(Spacer(1, 16))

    business_block = []
    business_name = invoice_data.get("business_name", "")
    if business_name:
        business_block.append(Paragraph(business_name, business_name_style))
    business_block.extend(_address_paragraphs(invoice_data.get("business_address", ""), normal))
    business_abn = invoice_data.get("business_abn", "")
    if business_abn:
        business_block.append(Paragraph(f"ABN: {format_abn(business_abn)}", normal))
    if not business_block:
        business_block.append(Paragraph("&nbsp;", normal))

    normal_right = ParagraphStyle("NormalRight", parent=normal, alignment=TA_RIGHT)
    label_style_right = ParagraphStyle("LabelRight", parent=label_style, alignment=TA_RIGHT)
    bill_to_lines = [Paragraph("BILL TO", label_style_right)]
    bill_to_lines.append(Paragraph(f"<b>{invoice_data['customer_name']}</b>", normal_right))
    bill_to_lines.extend(_address_paragraphs(invoice_data.get("customer_address", ""), normal_right))
    customer_abn = invoice_data.get("customer_abn", "")
    if customer_abn:
        bill_to_lines.append(Paragraph(f"ABN: {format_abn(customer_abn)}", normal_right))

    details_table = Table(
        [[business_block, bill_to_lines]],
        colWidths=[CONTENT_WIDTH * 0.55, CONTENT_WIDTH * 0.45],
    )
    details_table.setStyle(
        TableStyle(
            [
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 0),
                ("RIGHTPADDING", (0, 0), (-1, -1), 0),
            ]
        )
    )
    story.append(details_table)
    story.append(Spacer(1, 16))

    amount_ex_gst = invoice_data["amount_ex_gst"]
    gst_amount = invoice_data["gst_amount"]
    total_inc_gst = invoice_data["total_inc_gst"]
    taxable_ex_gst = invoice_data.get("taxable_ex_gst", amount_ex_gst)
    gst_free_ex_gst = invoice_data.get("gst_free_ex_gst", Decimal("0"))
    line_items = invoice_data.get("line_items")
    if not line_items:
        line_items = [
            {
                "description": invoice_data.get("description", ""),
                "quantity": Decimal("1"),
                "unit_amount_ex_gst": amount_ex_gst,
                "amount_ex_gst": amount_ex_gst,
                "gst_applicable": gst_amount > 0,
            }
        ]


    amt_right = ParagraphStyle("AmtRight", parent=normal, alignment=TA_RIGHT)
    qty_right = ParagraphStyle("QtyRight", parent=normal, alignment=TA_RIGHT)
    hdr_right = ParagraphStyle("HdrRight", parent=normal, alignment=TA_RIGHT)

    if gst_registered:
        line_data = [
            [
                Paragraph("<b>Description</b>", normal),
                Paragraph("<b>Qty</b>", qty_right),
                Paragraph("<b>Unit (ex GST)</b>", hdr_right),
                Paragraph("<b>GST</b>", hdr_right),
                Paragraph("<b>Amount (ex GST)</b>", hdr_right),
            ],
        ]
        for item in line_items:
            qty = item.get("quantity", Decimal("1"))
            unit = item.get("unit_amount_ex_gst", item["amount_ex_gst"])
            gst_applicable = item.get("gst_applicable", True)
            gst_label = "10%" if gst_applicable else "No"
            line_data.append(
                [
                    Paragraph(item["description"], normal),
                    Paragraph(format_qty(qty), qty_right),
                    Paragraph(format_money(unit), amt_right),
                    Paragraph(gst_label, amt_right),
                    Paragraph(format_money(item["amount_ex_gst"]), amt_right),
                ]
            )
        line_col_widths = [
            CONTENT_WIDTH * 0.38,
            CONTENT_WIDTH * 0.08,
            CONTENT_WIDTH * 0.18,
            CONTENT_WIDTH * 0.08,
            CONTENT_WIDTH * 0.18,
        ]
    else:
        line_data = [
            [
                Paragraph("<b>Description</b>", normal),
                Paragraph("<b>Qty</b>", qty_right),
                Paragraph("<b>Unit price</b>", hdr_right),
                Paragraph("<b>Amount</b>", hdr_right),
            ],
        ]
        for item in line_items:
            qty = item.get("quantity", Decimal("1"))
            unit = item.get("unit_amount_ex_gst", item["amount_ex_gst"])
            line_data.append(
                [
                    Paragraph(item["description"], normal),
                    Paragraph(format_qty(qty), qty_right),
                    Paragraph(format_money(unit), amt_right),
                    Paragraph(format_money(item["amount_ex_gst"]), amt_right),
                ]
            )
        line_col_widths = [
            CONTENT_WIDTH * 0.46,
            CONTENT_WIDTH * 0.10,
            CONTENT_WIDTH * 0.22,
            CONTENT_WIDTH * 0.22,
        ]

    line_table = Table(line_data, colWidths=line_col_widths)
    line_table.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, 0), LIGHT_GREY),
                ("BOX", (0, 0), (-1, -1), 0.5, BORDER_GREY),
                ("INNERGRID", (0, 0), (-1, -1), 0.5, BORDER_GREY),
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("TOPPADDING", (0, 0), (-1, -1), 8),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 8),
                ("LEFTPADDING", (0, 0), (-1, -1), 8),
                ("RIGHTPADDING", (0, 0), (-1, -1), 8),
            ]
        )
    )
    story.append(line_table)
    story.append(Spacer(1, 12))

    totals_data = []
    if gst_registered:
        if gst_free_ex_gst > 0 and taxable_ex_gst > 0:
            totals_data.append(
                [Paragraph("Taxable subtotal (ex GST)", total_label_style), Paragraph(format_money(taxable_ex_gst), total_value_style)]
            )
            totals_data.append(
                [Paragraph("GST-free subtotal (ex GST)", total_label_style), Paragraph(format_money(gst_free_ex_gst), total_value_style)]
            )
        totals_data.append(
            [Paragraph("Subtotal (ex GST)", total_label_style), Paragraph(format_money(amount_ex_gst), total_value_style)]
        )
        gst_label = "GST (10%)" if gst_amount > 0 else "GST not applicable"
        totals_data.append(
            [Paragraph(gst_label, total_label_style), Paragraph(format_money(gst_amount), total_value_style)]
        )
        totals_data.append(
            [Paragraph("TOTAL AUD", grand_total_label), Paragraph(format_money(total_inc_gst), grand_total_value)]
        )
    else:
        totals_data.append(
            [Paragraph("TOTAL AUD", grand_total_label), Paragraph(format_money(total_inc_gst), grand_total_value)]
        )
    totals_table = Table(totals_data, colWidths=[CONTENT_WIDTH * 0.72, CONTENT_WIDTH * 0.28])
    totals_table.setStyle(
        TableStyle(
            [
                ("LINEABOVE", (0, len(totals_data) - 1), (-1, len(totals_data) - 1), 1, ACCENT),
                ("TOPPADDING", (0, len(totals_data) - 1), (-1, len(totals_data) - 1), 8),
                ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
                ("LEFTPADDING", (0, 0), (-1, -1), 0),
                ("RIGHTPADDING", (0, 0), (-1, -1), 8),
            ]
        )
    )
    story.append(totals_table)
    story.append(Spacer(1, 20))

    account_name = invoice_data.get("account_name", "")
    bsb = invoice_data.get("bsb", "")
    acc = invoice_data.get("acc", "")
    if account_name or bsb or acc or due_date_fmt or payment_terms:
        payment_lines = [Paragraph("<b>How to pay</b>", ParagraphStyle("PayHdr", parent=normal, fontSize=11, spaceAfter=6))]
        if account_name:
            payment_lines.append(Paragraph(f"Account name: {account_name}", normal))
        if bsb:
            payment_lines.append(Paragraph(f"BSB: {format_bsb(bsb)}", normal))
        if acc:
            payment_lines.append(Paragraph(f"Account: {format_account(acc)}", normal))
        if due_date_fmt:
            terms_note = f" ({due_rule_label})" if due_rule_label else ""
            payment_lines.append(Paragraph(f"Payment due: {due_date_fmt}{terms_note}", normal))
        elif payment_terms:
            payment_lines.append(Paragraph(f"Payment terms: {payment_terms}", normal))
        payment_lines.append(Paragraph(f"Reference: Invoice #{invoice_number_display}", normal))

        payment_table = Table([[payment_lines]], colWidths=[CONTENT_WIDTH])
        payment_table.setStyle(
            TableStyle(
                [
                    ("BOX", (0, 0), (-1, -1), 1, BORDER_GREY),
                    ("BACKGROUND", (0, 0), (-1, -1), LIGHT_GREY),
                    ("TOPPADDING", (0, 0), (-1, -1), 12),
                    ("BOTTOMPADDING", (0, 0), (-1, -1), 12),
                    ("LEFTPADDING", (0, 0), (-1, -1), 12),
                    ("RIGHTPADDING", (0, 0), (-1, -1), 12),
                ]
            )
        )
        story.append(payment_table)

    comment = invoice_data.get("comment", "")
    if comment:
        story.append(Spacer(1, 12))
        story.append(Paragraph("<b>Notes</b>", label_style))
        story.append(Paragraph(comment, normal))

    work_photos = invoice_data.get("work_photos") or []
    if work_photos:
        story.append(Spacer(1, 16))
        story.append(Paragraph("<b>Work completed</b>", label_style))
        for photo_path in work_photos:
            if not photo_path or not os.path.isfile(photo_path):
                continue
            try:
                img = Image(photo_path)
                max_w = CONTENT_WIDTH
                max_h = 80 * mm
                iw, ih = float(img.imageWidth), float(img.imageHeight)
                scale = min(max_w / iw, max_h / ih, 1.0) if iw and ih else 1.0
                img.drawWidth = iw * scale
                img.drawHeight = ih * scale
                story.append(Spacer(1, 8))
                story.append(img)
            except Exception:
                continue

    doc.build(story)
    return filepath

File: account_api/worker/src/pdf/templates/classic.js
import { PDFDocument, StandardFonts, rgb } from "pdf-lib";
import {
  ACCENT,
  BORDER_GREY,
  CONTENT_WIDTH,
  HEADER_SLOT_FRACTION,
  LIGHT_GREY,
  MUTED,
  PAGE_HEIGHT,
  PAGE_MARGIN,
  PAGE_WIDTH,
  TEXT,
} from "../geometry.js";
import {
  formatAbn,
  formatAccount,
  formatAddressMultiline,
  formatBsb,
  formatDisplayDate,
  formatInvoiceNumber,
  formatMoney,
  formatQty,
} from "../format.js";

function dataUrlToBytes(dataUrl) {
  const raw = String(dataUrl || "");
  const comma = raw.indexOf(",");
  const b64 = comma >= 0 ? raw.slice(comma + 1) : raw;
  const bin = atob(b64);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

async function embedImage(doc, dataUrl) {
  if (!dataUrl) return null;
  const bytes = dataUrlToBytes(dataUrl);
  try {
    if (String(dataUrl).includes("image/png")) return await doc.embedPng(bytes);
    return await doc.embedJpg(bytes);
  } catch {
    try {
      return await doc.embedPng(bytes);
    } catch {
      return null;
    }
  }
}

function color(c) {
  return rgb(c.r, c.g, c.b);
}

/**
 * Classic AU tax-invoice layout (pdf-lib port of desktop classic.py).
 * @param {object} invoice
 * @param {object} business
 * @param {object} [customer]
 */
export async function renderClassicInvoice(invoice, business = {}, customer = {}) {
  const doc = await PDFDocument.create();
  const font = await doc.embedFont(StandardFonts.Helvetica);
  const bold = await doc.embedFont(StandardFonts.HelveticaBold);

  let page = doc.addPage([PAGE_WIDTH, PAGE_HEIGHT]);
  let y = PAGE_HEIGHT - PAGE_MARGIN;

  const ensure = (need) => {
    if (y < PAGE_MARGIN + need) {
      page = doc.addPage([PAGE_WIDTH, PAGE_HEIGHT]);
      y = PAGE_HEIGHT - PAGE_MARGIN;
    }
  };

  const drawText = (text, x, yy, size, useBold = false, col = TEXT, maxW = CONTENT_WIDTH) => {
    const f = useBold ? bold : font;
    let line = String(text || "");
    while (line && f.widthOfTextAtSize(line, size) > maxW) {
      line = line.slice(0, -1);
    }
    page.drawText(line, { x, y: yy, size, font: f, color: color(col) });
    return f.heightAtSize(size);
  };

  const drawRight = (text, rightX, yy, size, useBold = false, col = TEXT) => {
    const f = useBold ? bold : font;
    const tw = f.widthOfTextAtSize(String(text || ""), size);
    page.drawText(String(text || ""), {
      x: rightX - tw,
      y: yy,
      size,
      font: f,
      color: color(col),
    });
  };

  const gstRegistered = Boolean(
    invoice.gst_registered ?? (Number(invoice.gst_amount || 0) > 0 || business.gst_registered)
  );
  const title = gstRegistered ? "Tax Invoice" : "Invoice";
  const invNum = formatInvoiceNumber(invoice.invoice_number);
  const dueFmt = invoice.due_date_fmt || formatDisplayDate(invoice.due_date);
  const leftW = CONTENT_WIDTH * HEADER_SLOT_FRACTION;
  const rightW = CONTENT_WIDTH * (1 - HEADER_SLOT_FRACTION);
  const leftX = PAGE_MARGIN;
  const rightX = PAGE_MARGIN + leftW;

  // Header: logo | meta
  let headerBottom = y;
  const metaLines = [
    { text: title, size: 18, bold: true },
    { text: `Invoice # ${invNum}`, size: 10, bold: false },
    { text: `Date ${formatDisplayDate(invoice.invoice_date)}`, size: 10, bold: false },
  ];
  if (dueFmt) metaLines.push({ text: `Due ${dueFmt}`, size: 10, bold: false });

  let metaY = y - 4;
  for (const line of metaLines) {
    drawRight(line.text, PAGE_MARGIN + CONTENT_WIDTH, metaY - line.size, line.size, line.bold, ACCENT);
    metaY -= line.size + (line.bold ? 10 : 4);
  }
  const metaHeight = y - metaY;

  let logoHeight = 0;
  const logoEnabled = business.logo_enabled !== false && Boolean(business.logo_b64);
  if (logoEnabled && business.logo_b64) {
    const img = await embedImage(doc, business.logo_b64);
    if (img) {
      const maxW = leftW - 4;
      const scale = Math.min(1, maxW / img.width, (metaHeight || 80) / img.height * 1.2);
      const w = img.width * Math.min(scale, maxW / img.width);
      const h = img.height * (w / img.width);
      logoHeight = h;
      page.drawImage(img, { x: leftX, y: y - h, width: w, height: h });
    }
  }

  y -= Math.max(metaHeight, logoHeight, 48) + 8;

  // Divider
  ensure(20);
  page.drawRectangle({
    x: PAGE_MARGIN,
    y: y - 2,
    width: CONTENT_WIDTH,
    height: 2,
    color: color(ACCENT),
  });
  y -= 18;

  // Parties
  ensure(80);
  const bizName = business.business_name || invoice.business_name || "";
  const bizLines = [bizName, ...formatAddressMultiline(business)];
  const abn = business.business_abn || business.abn || "";
  if (abn) bizLines.push(`ABN: ${formatAbn(abn)}`);

  let partyY = y;
  for (let i = 0; i < bizLines.length; i++) {
    if (!bizLines[i]) continue;
    drawText(bizLines[i], leftX, partyY - 11, i === 0 ? 12 : 10, i === 0);
    partyY -= i === 0 ? 16 : 13;
  }

  const cust = customer || {};
  const billLines = ["BILL TO", String(invoice.customer_name || ""), ...formatAddressMultiline(cust)];
  const custAbn = cust.abn || cust.business_abn || "";
  if (custAbn) billLines.push(`ABN: ${formatAbn(custAbn)}`);

  let billY = y;
  for (let i = 0; i < billLines.length; i++) {
    if (!billLines[i]) continue;
    const size = i === 0 ? 9 : 10;
    const isBold = i === 1;
    drawRight(billLines[i], PAGE_MARGIN + CONTENT_WIDTH, billY - size, size, isBold, i === 0 ? MUTED : TEXT);
    billY -= size + 3;
  }

  y = Math.min(partyY, billY) - 16;

  // Line items table
  const items = invoice.line_items?.length
    ? invoice.line_items
    : invoice.description
      ? [
          {
            description: invoice.description,
            quantity: 1,
            unit_amount_ex_gst: invoice.amount_ex_gst,
            amount_ex_gst: invoice.amount_ex_gst,
            gst_applicable: Number(invoice.gst_amount || 0) > 0,
          },
        ]
      : [];

  const cols = gstRegistered
    ? [
        { key: "desc", label: "Description", w: CONTENT_WIDTH * (38 / 90), align: "left" },
        { key: "qty", label: "Qty", w: CONTENT_WIDTH * (8 / 90), align: "right" },
        { key: "unit", label: "Unit (ex GST)", w: CONTENT_WIDTH * (18 / 90), align: "right" },
        { key: "gst", label: "GST", w: CONTENT_WIDTH * (8 / 90), align: "right" },
        { key: "amt", label: "Amount (ex GST)", w: CONTENT_WIDTH * (18 / 90), align: "right" },
      ]
    : [
        { key: "desc", label: "Description", w: CONTENT_WIDTH * 0.46, align: "left" },
        { key: "qty", label: "Qty", w: CONTENT_WIDTH * 0.1, align: "right" },
        { key: "unit", label: "Unit price", w: CONTENT_WIDTH * 0.22, align: "right" },
        { key: "amt", label: "Amount", w: CONTENT_WIDTH * 0.22, align: "right" },
      ];

  const rowH = 22;
  ensure(rowH * (items.length + 2) + 40);

  // Header row
  page.drawRectangle({
    x: PAGE_MARGIN,
    y: y - rowH,
    width: CONTENT_WIDTH,
    height: rowH,
    color: color(LIGHT_GREY),
    borderColor: color(BORDER_GREY),
    borderWidth: 0.5,
  });
  let cx = PAGE_MARGIN;
  for (const col of cols) {
    const labelY = y - 15;
    if (col.align === "right") drawRight(col.label, cx + col.w - 6, labelY, 8, true);
    else drawText(col.label, cx + 6, labelY, 8, true);
    cx += col.w;
  }
  y -= rowH;

  for (const item of items) {
    ensure(rowH + 10);
    page.drawRectangle({
      x: PAGE_MARGIN,
      y: y - rowH,
      width: CONTENT_WIDTH,
      height: rowH,
      borderColor: color(BORDER_GREY),
      borderWidth: 0.5,
    });
    const qty = item.quantity ?? 1;
    const unit = item.unit_amount_ex_gst ?? item.amount_ex_gst;
    const amt = item.amount_ex_gst ?? unit;
    const gstLabel = item.gst_applicable === false ? "No" : "10%";
    const cells = gstRegistered
      ? {
          desc: item.description || "",
          qty: formatQty(qty),
          unit: formatMoney(unit),
          gst: gstLabel,
          amt: formatMoney(amt),
        }
      : {
          desc: item.description || "",
          qty: formatQty(qty),
          unit: formatMoney(unit),
          amt: formatMoney(amt),
        };
    cx = PAGE_MARGIN;
    for (const col of cols) {
      const labelY = y - 15;
      const val = cells[col.key] || "";
      if (col.align === "right") drawRight(val, cx + col.w - 6, labelY, 9);
      else drawText(val, cx + 6, labelY, 9, false, TEXT, col.w - 12);
      cx += col.w;
    }
    y -= rowH;
  }

  y -= 12;

  // Totals
  const amountEx = Number(invoice.amount_ex_gst || 0);
  const gstAmt = Number(invoice.gst_amount || 0);
  const total = Number(invoice.total_inc_gst || 0);
  const taxable = Number(invoice.taxable_ex_gst ?? amountEx);
  const gstFree = Number(invoice.gst_free_ex_gst || 0);

  const totals = [];
  if (gstRegistered) {
    if (gstFree > 0 && taxable > 0) {
      totals.push(["Taxable subtotal (ex GST)", formatMoney(taxable)]);
      totals.push(["GST-free subtotal (ex GST)", formatMoney(gstFree)]);
    }
    totals.push(["Subtotal (ex GST)", formatMoney(amountEx)]);
    totals.push([gstAmt > 0 ? "GST (10%)" : "GST not applicable", formatMoney(gstAmt)]);
    totals.push(["TOTAL AUD", formatMoney(total)]);
  } else {
    totals.push(["TOTAL AUD", formatMoney(total)]);
  }

  ensure(totals.length * 16 + 20);
  for (let i = 0; i < totals.length; i++) {
    const [label, val] = totals[i];
    const isGrand = i === totals.length - 1;
    if (isGrand) {
      // Clear previous totals row before the TOTAL rule (GST-on only; GST-off has one row).
      if (i > 0) y -= 8;
      page.drawLine({
        start: { x: PAGE_MARGIN + CONTENT_WIDTH * 0.5, y: y + 4 },
        end: { x: PAGE_MARGIN + CONTENT_WIDTH, y: y + 4 },
        thickness: 1,
        color: color(ACCENT),
      });
      y -= 4;
    }
    drawRight(label, PAGE_MARGIN + CONTENT_WIDTH * 0.72, y - 11, isGrand ? 11 : 10, true);
    drawRight(val, PAGE_MARGIN + CONTENT_WIDTH, y - 11, isGrand ? 11 : 10, isGrand);
    y -= isGrand ? 18 : 14;
  }

  y -= 12;

  // How to pay
  const accountName = business.account_name || "";
  const bsb = business.bsb || "";
  const acc = business.acc || "";
  if (accountName || bsb || acc || dueFmt) {
    ensure(100);
    const boxH = 14 * (4 + (dueFmt ? 1 : 0) + 1) + 16;
    page.drawRectangle({
      x: PAGE_MARGIN,
      y: y - boxH,
      width: CONTENT_WIDTH,
      height: boxH,
      color: color(LIGHT_GREY),
      borderColor: color(BORDER_GREY),
      borderWidth: 1,
    });
    let py = y - 16;
    drawText("How to pay", PAGE_MARGIN + 12, py, 11, true);
    py -= 16;
    if (accountName) {
      drawText(`Account name: ${accountName}`, PAGE_MARGIN + 12, py, 10);
      py -= 14;
    }
    if (bsb) {
      drawText(`BSB: ${formatBsb(bsb)}`, PAGE_MARGIN + 12, py, 10);
      py -= 14;
    }
    if (acc) {
      drawText(`Account: ${formatAccount(acc)}`, PAGE_MARGIN + 12, py, 10);
      py -= 14;
    }
    if (dueFmt) {
      drawText(`Payment due: ${dueFmt}`, PAGE_MARGIN + 12, py, 10);
      py -= 14;
    }
    drawText(`Reference: Invoice #${invNum}`, PAGE_MARGIN + 12, py, 10);
    y -= boxH + 12;
  }

  if (invoice.comment) {
    ensure(40);
    drawText("Notes", PAGE_MARGIN, y - 10, 9, true, MUTED);
    y -= 16;
    drawText(String(invoice.comment), PAGE_MARGIN, y - 10, 10);
    y -= 20;
  }

  const photos = invoice.work_photos_b64 || [];
  if (photos.length) {
    ensure(30);
    drawText("Work completed", PAGE_MARGIN, y - 10, 9, true, MUTED);
    y -= 16;
    for (const src of photos.slice(0, 6)) {
      const img = await embedImage(doc, src);
      if (!img) continue;
      const maxW = CONTENT_WIDTH;
      const maxH = 226; // ~80mm
      const scale = Math.min(1, maxW / img.width, maxH / img.height);
      const w = img.width * scale;
      const h = img.height * scale;
      ensure(h + 16);
      page.drawImage(img, { x: PAGE_MARGIN, y: y - h, width: w, height: h });
      y -= h + 12;
    }
  }

  return doc.save();
}
